File: src/main/java/gg/mineral/practice/listeners/InventoryListener.java
package gg.mineral.practice.listeners;

import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.Material;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.event.player.PlayerDropItemEvent;
import org.bukkit.event.player.PlayerPickupItemEvent;

import gg.mineral.practice.PracticePlugin;
import gg.mineral.practice.entity.PlayerStatus;
import gg.mineral.practice.inventory.Interaction;
import gg.mineral.practice.managers.ProfileManager;
import lombok.val;

public class InventoryListener implements Listener {

	@EventHandler
	public void onInventoryClick(InventoryClickEvent e) {
		val profile = ProfileManager.getOrCreateProfile((org.bukkit.entity.Player) e.getWhoClicked());
		val menu = profile.getOpenMenu();

		boolean canClick = profile.getPlayer().isOp() && profile.getPlayer().getGameMode().equals(GameMode.CREATIVE);

		val clickType = e.getClick();

		e.setCancelled(e.getCurrentItem() == null ? false : e.getCurrentItem().getType() == Material.TNT);

		if (profile.getInventory().isInventoryClickCancelled())
			e.setCancelled(!canClick);

		if (menu == null)
			return;

		if (e.getClickedInventory() == null || !e.getClickedInventory().equals(e.getView().getTopInventory()))
			return;

		if (e.getInventory() == null || !e.getInventory().equals(menu.getInventory()))
			return;

		if (e.getSlot() < e.getView().getTopInventory().getSize())
			e.setCancelled(
					menu.isClickCancelled());

		val predicate = menu.getTask(e.getSlot());

		if (predicate == null)
			return;

		predicate.accept(new Interaction(profile, clickType));
	}

	@EventHandler
	public void onInventoryClose(InventoryCloseEvent e) {
		val profile = ProfileManager.getOrCreateProfile((org.bukkit.entity.Player) e.getPlayer());
		val oldMenu = profile.getOpenMenu();
		profile.setOpenMenu(null);

		if (oldMenu != null && !oldMenu.isClosed() && e.getInventory().equals(oldMenu.getInventory())) {
			oldMenu.setClosed(true);
			Bukkit.getScheduler().scheduleSyncDelayedTask(PracticePlugin.INSTANCE, () -> oldMenu.onClose(), 1);
		}
	}

	@EventHandler
	public void onPlayerDropItem(PlayerDropItemEvent e) {
		val player = e.getPlayer();
		boolean canDrop = player.isOp() && player.getGameMode().equals(GameMode.CREATIVE);

		val profile = ProfileManager.getProfile(player);

		if (profile != null) {
			if (profile.getPlayerStatus() == PlayerStatus.KIT_CREATOR
					|| profile.getPlayerStatus() == PlayerStatus.KIT_EDITOR) {
				e.setCancelled(false);
				Bukkit.getScheduler().runTaskLater(PracticePlugin.INSTANCE, () -> e.getItemDrop().remove(), 20L);
				return;
			}

			val match = profile.getMatch();

			if (match != null) {
				match.getItemRemovalQueue().add(e.getItemDrop());
				return;
			}
		}

		e.setCancelled(!canDrop);

	}

	@EventHandler
	public void onPlayerPickupItem(PlayerPickupItemEvent e) {
		boolean canPickup = e.getPlayer().isOp() && e.getPlayer().getGameMode().equals(GameMode.CREATIVE);

		if (canPickup)
			return;

		e.setCancelled(ProfileManager
				.getProfile(e.getPlayer().getUniqueId(),
						p -> p.getPlayerStatus() == PlayerStatus.FIGHTING) == null);
	}
}

File: src/main/kotlin/gg/mineral/practice/listeners/InventoryListener.kt
package gg.mineral.practice.listeners

import gg.mineral.practice.PracticePlugin
import gg.mineral.practice.entity.PlayerStatus
import gg.mineral.practice.entity.Profile
import gg.mineral.practice.inventory.Interaction
import gg.mineral.practice.managers.ProfileManager.getOrCreateProfile
import gg.mineral.practice.managers.ProfileManager.getProfile
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.player.PlayerDropItemEvent
import org.bukkit.event.player.PlayerPickupItemEvent

class InventoryListener : Listener {
    @EventHandler
    fun onInventoryClick(e: InventoryClickEvent) {
        val profile = getOrCreateProfile(e.whoClicked as Player)
        val menu = profile.openMenu

        val canClick = profile.player.isOp && profile.player.gameMode == GameMode.CREATIVE

        val clickType = e.click

        e.isCancelled = e.currentItem != null && e.currentItem.type == Material.TNT

        if (profile.inventory.inventoryClickCancelled) e.isCancelled = !canClick

        if (menu == null) return

        if (e.clickedInventory == null || e.clickedInventory != e.view.topInventory) return

        if (e.inventory == null || e.inventory != menu.inventory) return

        if (e.slot < e.view.topInventory.size) e.isCancelled = menu.clickCancelled

        val predicate = menu.getTask(e.slot) ?: return

        predicate.accept(Interaction(profile, clickType))
    }

    @EventHandler
    fun onInventoryClose(e: InventoryCloseEvent) {
        val profile = getOrCreateProfile(e.player as Player)
        val oldMenu = profile.openMenu
        profile.openMenu = null

        if (oldMenu != null && !oldMenu.closed && e.inventory == oldMenu.inventory) {
            oldMenu.closed = true
            Bukkit.getScheduler().scheduleSyncDelayedTask(
                PracticePlugin.INSTANCE,
                { oldMenu.onClose() }, 1
            )
        }
    }

    @EventHandler
    fun onPlayerDropItem(e: PlayerDropItemEvent) {
        val player = e.player
        val canDrop = player.isOp && player.gameMode == GameMode.CREATIVE

        val profile = getProfile(player)

        if (profile != null) {
            if (profile.playerStatus === PlayerStatus.KIT_CREATOR
                || profile.playerStatus === PlayerStatus.KIT_EDITOR
            ) {
                e.isCancelled = false
                Bukkit.getScheduler().runTaskLater(
                    PracticePlugin.INSTANCE,
                    { e.itemDrop.remove() }, 20L
                )
                return
            }

            val match = profile.match

            if (match != null) {
                match.itemRemovalQueue.add(e.itemDrop)
                return
            }
        }

        e.isCancelled = !canDrop
    }

    @EventHandler
    fun onPlayerPickupItem(e: PlayerPickupItemEvent) {
        val canPickup = e.player.isOp && e.player.gameMode == GameMode.CREATIVE

        if (canPickup) return

        e.isCancelled = getProfile(
            e.player.uniqueId
        ) { p: Profile -> p.playerStatus === PlayerStatus.FIGHTING } == null
    }
}
